File: overlay/python/hydra-kernel/src/main/python/hydra/dsl/meta/lib/optionals.py
"""Phantom-typed term DSL for the hydra.lib.optionals library."""

from hydra.typed import TypedTerm
from hydra.dsl.meta.phantoms import lib_primitive as primitive, lib_primitive1 as primitive1, lib_primitive2 as primitive2, lib_primitive3 as primitive3


def apply(f: TypedTerm, x: TypedTerm) -> TypedTerm:
    """Apply a function to an argument."""
    return primitive2(f, x)


def bind(x: TypedTerm, f: TypedTerm) -> TypedTerm:
    """Chain operations on optional values, handling Nothing cases automatically."""
    return primitive2(x, f)


def cases(m: TypedTerm, n: TypedTerm, j: TypedTerm) -> TypedTerm:
    """Case analysis on an optional value (scrutinee-first): cases(m, ifNothing, ifJust)."""
    return primitive3(m, n, j)


def cat(xs: TypedTerm) -> TypedTerm:
    """Filter out Nothing values from a list."""
    return primitive1(xs)


def compose(f: TypedTerm, g: TypedTerm) -> TypedTerm:
    """Compose two functions."""
    return primitive2(f, g)


def from_just(x: TypedTerm) -> TypedTerm:
    """Extract value from Maybe, assuming it's Just (unsafe)."""
    return primitive1(x)


def from_optional(default: TypedTerm, x: TypedTerm) -> TypedTerm:
    """Get a value from an optional value, or return a default value."""
    return primitive2(default, x)


def is_given(x: TypedTerm) -> TypedTerm:
    """Check if a value is Just."""
    return primitive1(x)


def is_none(x: TypedTerm) -> TypedTerm:
    """Check if a value is Nothing."""
    return primitive1(x)


def map(f: TypedTerm, x: TypedTerm) -> TypedTerm:
    """Map a function over an optional value."""
    return primitive2(f, x)


def map_optional(f: TypedTerm, xs: TypedTerm) -> TypedTerm:
    """Map a function over a list and collect Just results."""
    return primitive2(f, xs)


def pure(x: TypedTerm) -> TypedTerm:
    """Lift a value into the Maybe type."""
    return primitive1(x)


def to_list(x: TypedTerm) -> TypedTerm:
    """Convert a Maybe to a list: Just x becomes [x], Nothing becomes []."""
    return primitive1(x)

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Dsl/Meta/Lib/Optionals.hs
{-# LANGUAGE FlexibleContexts #-}

-- | Phantom-typed term DSL for the hydra.lib.optionals library

module Hydra.Dsl.Meta.Lib.Optionals where

import Hydra.Typed
import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import qualified Hydra.Dsl.Prims as Prims
import qualified Hydra.Lib.Optionals as DefOptionals


-- | Apply a function to an argument (applicative).
apply :: TypedTerm (Maybe (a -> b)) -> TypedTerm (Maybe a) -> TypedTerm (Maybe b)
apply = primitive2 DefOptionals.apply

-- | Chain operations on optional values, handling absent cases automatically.
bind :: TypedTerm (Maybe a) -> TypedTerm (a -> Maybe b) -> TypedTerm (Maybe b)
bind = primitive2 DefOptionals.bind

-- | Handle an optional value with the optional value as the first argument.
-- The default and function arguments accept anything coercible to a term (AsTerm),
-- mirroring the ergonomics of the former 'maybe' eliminator.
cases :: (AsTerm t1 b, AsTerm t2 (a -> b)) => TypedTerm (Maybe a) -> t1 -> t2 -> TypedTerm b
cases m def f = primitive3 DefOptionals.cases m (asTerm def) (asTerm f)

-- | Filter out absent values from a list.
cat :: TypedTerm [Maybe a] -> TypedTerm [a]
cat = primitive1 DefOptionals.cat

-- | Compose two optional-returning functions (Kleisli composition).
compose :: TypedTerm (a -> Maybe b) -> TypedTerm (b -> Maybe c) -> TypedTerm (a -> Maybe c)
compose = primitive2 DefOptionals.compose

-- | Get a value from an optional value, or return a default value.
fromOptional :: TypedTerm a -> TypedTerm (Maybe a) -> TypedTerm a
fromOptional = primitive2 DefOptionals.fromOptional

-- | Check if a value is present.
isGiven :: TypedTerm (Maybe a) -> TypedTerm Bool
isGiven = primitive1 DefOptionals.isGiven

-- | Check if a value is absent.
isNone :: TypedTerm (Maybe a) -> TypedTerm Bool
isNone = primitive1 DefOptionals.isNone

-- | Map a function over an optional value.
map :: AsTerm f (a -> b) => f -> TypedTerm (Maybe a) -> TypedTerm (Maybe b)
map f = primitive2 DefOptionals.map (asTerm f)

-- | Map a function over a list and collect present results.
mapOptional :: TypedTerm (a -> Maybe b) -> TypedTerm [a] -> TypedTerm [b]
mapOptional = primitive2 DefOptionals.mapOptional

-- | Lift a value into the optional type.
pure :: TypedTerm a -> TypedTerm (Maybe a)
pure = primitive1 DefOptionals.pure

-- | Convert an optional value to a list: a present value becomes [x], an absent value becomes [].
toList :: TypedTerm (Maybe a) -> TypedTerm [a]
toList = primitive1 DefOptionals.toList
